include/igl/opengl/ViewerData.cpp

```diff
@@ -795,4 +797,101 @@ IGL_INLINE void igl::opengl::ViewerData::updateGL(
+    meshgl.vid_label_char_vbo.resize(totalVertCharacters,1);
+    meshgl.vid_label_offset_vbo.resize(totalVertCharacters,1);
+    meshgl.vid_label_indices_vbo.resize(totalVertCharacters,1);
+    Eigen::MatrixXd normalized = V_normals.normalized();
+    int charCount=0;
+    for (int v=0; v<V.rows();++v)
+    {
+      std::string vertName = std::to_string(v);
+      for(int c=0; c<vertName.length(); c++)
+      {
+        meshgl.vid_label_pos_vbo.row(charCount) = ((normalized*0.1).row(v) + V.row(v)).cast<float>();
+        meshgl.vid_label_char_vbo(charCount) = (float)(vertName.at(c));
+        meshgl.vid_label_offset_vbo(charCount) = c;
+        meshgl.vid_label_indices_vbo(charCount) = charCount;
+        charCount++;
+      }
+    }
+  }
+
+  if (meshgl.dirty & MeshGL::DIRTY_FID_LABELS)
+  {
+    std::string faceCount = std::to_string(F.rows());
+    int totalFaceCharacters = 0;
+    int i=1;
+    while(i<faceCount.length())
+    {
+      totalFaceCharacters += (i)*9*pow(10,i-1); i++;
+    }
+    totalFaceCharacters += (i*(F.rows() - pow(10,i-1)));
+    totalFaceCharacters += 1;
+    meshgl.fid_label_pos_vbo.resize(totalFaceCharacters,3);
+    meshgl.fid_label_char_vbo.resize(totalFaceCharacters,1);
+    meshgl.fid_label_offset_vbo.resize(totalFaceCharacters,1);
+    meshgl.fid_label_indices_vbo.resize(totalFaceCharacters,1);
+    Eigen::MatrixXd faceNormals = F_normals.normalized();
+    int charCount=0;
+    for (int f=0; f<F.rows();++f)
+    {
+      std::string faceName = std::to_string(f);
+      for(int c=0; c<faceName.length(); c++)
+      {
+        meshgl.fid_label_pos_vbo.row(charCount) =  V.row(F.row(f)(0)).cast<float>();
+        meshgl.fid_label_pos_vbo.row(charCount) += V.row(F.row(f)(1)).cast<float>();
+        meshgl.fid_label_pos_vbo.row(charCount) += V.row(F.row(f)(2)).cast<float>();
+        meshgl.fid_label_pos_vbo.row(charCount) /= 3.;
+        meshgl.fid_label_pos_vbo.row(charCount) = (faceNormals*0.05).row(f).cast<float>() + meshgl.fid_label_pos_vbo.row(charCount);
+
+        meshgl.fid_label_char_vbo(charCount) = (float)(faceName.at(c));
+        meshgl.fid_label_offset_vbo(charCount) = c;
+        meshgl.fid_label_indices_vbo(charCount) = charCount;
+        charCount++;
+      }
+    }
+  }
+
+  if (meshgl.dirty & MeshGL::DIRTY_EXTRA_LABELS)
+  {
+    if(data.labels_positions.rows()>0)
+    {
+      int numCustomChars = 0;
+      for(int i=0; i<labels_positions.rows(); i++)
+      {
+        numCustomChars += labels_strings.at(i).length();
+      }
+      meshgl.extra_label_pos_vbo.resize(numCustomChars, 3);
+      meshgl.extra_label_char_vbo.resize(numCustomChars, 1);
+      meshgl.extra_label_offset_vbo.resize(numCustomChars, 1);
+      meshgl.extra_label_indices_vbo.resize(numCustomChars, 1);
+      int charCount=0;
+      assert(labels_strings.size() == labels_positions.rows());
+      for(int s=0; s<labels_strings.size(); s++)
+      {
+        std::string temp = labels_strings.at(s);
```

Comment:
Do not use `std::string` here, since it will copy the data. You can use `const auto & label = ...` if you want.